graylog2-web-interface/src/views/components/aggregationbuilder/HeatmapVisualizationConfiguration.tsx

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+import React, { useCallback, useState } from 'react';
+import styled from 'styled-components';
+
+import { Checkbox } from 'components/graylog';
+import { Input } from 'components/bootstrap';
+import HeatmapVisualizationConfig, { COLORSCALES } from 'views/logic/aggregationbuilder/visualizations/HeatmapVisualizationConfig';
+
+import Select from '../Select';
+
+const StyledInput = styled(Input)`
+  > label {
+    font-weight: normal;
+  }
+`;
+
+type Props = {
+  onChange: (config: HeatmapVisualizationConfig) => void,
+  config: HeatmapVisualizationConfig,
+};
+
+const _makeOption = (value) => ({ label: value, value });
+const colorScalesOptions = COLORSCALES.map(_makeOption);
+
+type Error = {
+  zmin?: string,
+  zmax?: string,
+  defaultValue?: string,
+}
+
+const _validateConfig = (config: HeatmapVisualizationConfig, setErrors): boolean => {
+  const { zMax, zMin, defaultValue } = config;
+  const errors = {} as Error;
+
+  if (zMin || zMax) {
+    if (zMin >= zMax) {
+      errors.zmin = 'Min is bigger than Max';
+    }
+
+    if (zMax <= zMin) {
+      errors.zmin = 'Max is smaller than Min';
+    }
+
+    if (defaultValue) {
+      if (defaultValue > zMax || defaultValue < zMin) {
+        errors.defaultValue = 'Default Value is out of range from Min and Max';
+      }
+    }
+  }
+
+  setErrors(errors);
+
+  return Object.keys(errors).length <= 0;
+};
+
+const _parseEvent = (event) => {
+  const { value } = event.target;
+
+  if (value === undefined) {
+    return undefined;
+  }
+
+  return parseFloat(value);
+};
+
+const HeatmapVisualizationConfiguration = ({ config = HeatmapVisualizationConfig.empty(), onChange: onChangeFunc }: Props) => {
+  const [errors, setErrors] = useState<Error>({});
+  const onChange = useCallback((newConfig) => {
+    if (_validateConfig(newConfig, setErrors)) {
+      onChangeFunc(newConfig);
+    }
+  }, [onChangeFunc]);
+  const _onColorScaleChange = useCallback(({ value }) => onChange(config.toBuilder().colorScale(value).build()), [config, onChange]);
```

Comment:
We could clean this up a bit by introducing something like this:

```suggestion
  const modifyConfig = useCallback((fn: (builder: HeatmapVisualizationConfigBuilder) => HeatmapVisualizationConfigBuilder) => {
    onChange(fn(config.toBuilder()).build());
  }, [onChange, config]);
  const _onColorScaleChange = useCallback(({ value }) => modifyConfig((builder) => builder.colorScale(value)), [modifyConfig]);
```

The next five callbacks could then reuse `modifyConfig` as well. It would require exporting the type from the `Builder` class in `HeatmapVisualizationConfig` as `HeatmapVisualizationConfigBuilder` though.